### ISSUE TITLE
EXCEPTION_ACCESS_VIOLATION in C2 CompilerThread on Java 17

### ISSUE THREAD

**Issue by KeesVo** [adoptium/adoptium-support#437]
## Summary

Following a JVM crash, the following message was given:

A fatal error has been detected by the Java Runtime Environment:

 EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffeace5cce2, pid=33692, tid=27488

JRE version: OpenJDK Runtime Environment Temurin-17.0.1+12 (17.0.1+12) (build 17.0.1+12)
Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (17.0.1+12, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)
Problematic frame:
V  [jvm.dll+0x57cce2]

---------------  S U M M A R Y ------------

Command Line: -Dorg.gradle.internal.worker.tmpdir=C:\Git\repo\Implementation\Projects\WG\impl\WGMain\build\tmp\test\work -Dorg.gradle.native=false --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED -Xmx4G -Dfile.encoding=windows-1252 -Duser.country=US -Duser.language=en -Duser.variant -ea worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 1'

Host: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz, 12 cores, 63G,  Windows 10 , 64 bit Build 19041 (10.0.19041.1348)
Time: Tue Jan 25 16:29:20 2022 W. Europe Standard Time elapsed time: 116.196137 seconds (0d 0h 1m 56s)

---------------  T H R E A D  ---------------

Current thread (0x00000202041b9ee0):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=27488, stack(0x0000002107100000,0x0000002107200000)]

Current CompileTask:
C2: 116196 19158       4       javafx.scene.control.TableView$TableViewArrayListSelectionModel::updateSelection (568 bytes)

Stack: [0x0000002107100000,0x0000002107200000],  sp=0x00000021071fa6e0,  free space=1001k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [jvm.dll+0x57cce2]
V  [jvm.dll+0x57c9f4]
V  [jvm.dll+0x57bc68]
V  [jvm.dll+0x22022d]
V  [jvm.dll+0x22060f]
V  [jvm.dll+0x219e6a]
V  [jvm.dll+0x217345]
V  [jvm.dll+0x1a355d]
V  [jvm.dll+0x226e38]
V  [jvm.dll+0x224f7c]
V  [jvm.dll+0x7e3f5b]
V  [jvm.dll+0x7de51a]
V  [jvm.dll+0x673305]
C  [ucrtbase.dll+0x21bb2]
C  [KERNEL32.DLL+0x17034]
C  [ntdll.dll+0x52651]

It seems the exception occurs upon C2 compilation of a specific javafx class method.
We have encountered the same issue, referencing the same javafx method and identical call stack, multiple times now.
The exception does not (seem to) occur on Java 11 (openjdk version "11.0.4+11" 2019-07-16).

Further logging can be found here:
[hs_err_pid33692.log](https://github.com/adoptium/adoptium-support/files/7943245/hs_err_pid33692.log)

## Steps to reproduce

Unfortunately, we cannot reproduce this issue reliably. We have tried several methods to try and trigger this exception on purpose, but it only seems to occur under specific conditions which we have not (yet) isolated. Using -XX:+PrintCompilation we see that the specific method mentioned in the log (javafx.scene.control.TableView$TableViewArrayListSelectionModel::updateSelection) is compiled and invalidated several times, but we don't know how to trigger the exception.

Since we cannot reproduce the issue deliberately, we can't be sure the issue is not present in Java 11 (11.0.4+11).

## Triaging info

Java (-)version:

openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment Temurin-17.0.1+12 (build 17.0.1+12)
OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (build 17.0.1+12, mixed mode)

Javafx version 15 used.

For system info, see also log file:

---------------  S Y S T E M  ---------------

OS:
 Windows 10 , 64 bit Build 19041 (10.0.19041.1348)
OS uptime: 5 days 0:16 hours
Hyper-V role detected

CPU: total 12 (initial active 12) (6 cores per cpu, 2 threads per core) family 6 model 158 stepping 10 microcode 0xca, cx8, cmov, fxsr, ht, mmx, 3dnowpref, sse, sse2, sse3, ssse3, sse4.1, sse4.2, popcnt, lzcnt, tsc, tscinvbit, avx, avx2, aes, erms, clmul, bmi1, bmi2, adx, fma, vzeroupper, clflush, clflushopt, hv

Memory: 4k page, system-wide physical 65382M (43118M free)
TotalPageFile size 75110M (AvailPageFile size 50225M)
current process WorkingSet (physical memory assigned to process): 2208M, peak: 2217M
current process commit charge ("private bytes"): 2476M, peak: 2484M

vm_info: OpenJDK 64-Bit Server VM (17.0.1+12) for windows-amd64 JRE (17.0.1+12), built on Oct 20 2021 07:42:55 by "" with MS VC++ 16.7 (VS2019)

**Comment by karianna** [comment]
17.0.2 will be released over the next few days - can you try that then and post the results back here?

**Comment by xupwup** [comment]
I've managed to trigger the same crash on 17.0.2:

JRE version: OpenJDK Runtime Environment Temurin-17.0.2+8 (17.0.2+8) (build 17.0.2+8)
Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.2+8 (17.0.2+8, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)

[replay_pid34972.log](https://github.com/adoptium/adoptium-support/files/7957023/replay_pid34972.log)
[hs_err_pid34972.log](https://github.com/adoptium/adoptium-support/files/7957015/hs_err_pid34972.log)

---------------  S U M M A R Y ------------

Command Line: -Dorg.gradle.internal.worker.tmpdir=C:\Git\dev_acp-hlc-emulation\Implementation\Projects\ABC\impl\ABCMain\build\tmp\test\work -Dorg.gradle.native=false --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED -Xmx4G -Dfile.encoding=windows-1252 -Duser.country=US -Duser.language=en -Duser.variant -ea worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 1'

Host: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz, 12 cores, 31G,  Windows 10 , 64 bit Build 19041 (10.0.19041.1466)
Time: Fri Jan 28 09:27:35 2022 W. Europe Standard Time elapsed time: 145.097007 seconds (0d 0h 2m 25s)

---------------  T H R E A D  ---------------

Current thread (0x0000019ca81b0580):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=26572, stack(0x0000000c36500000,0x0000000c36600000)]


Current CompileTask:
C2: 145097 22981       4       javafx.scene.control.TableView$TableViewArrayListSelectionModel::updateSelection (568 bytes)

Stack: [0x0000000c36500000,0x0000000c36600000],  sp=0x0000000c365fa6c0,  free space=1001k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [jvm.dll+0x57d9f2]
V  [jvm.dll+0x57d704]

**Comment by karianna** [comment]
@xupwup Looks like this is a OpenJFX control originating bug, have you reported it over at the OpenJFX project?

**Comment by xupwup** [comment]
It looks like JavaFX's java code (or rather it's bytecode) crashes the C2 jit compiler. Doesn't that mean it has to be a jvm bug? If it crashed in native code (in a javafx dll) then I'd agree with you, but this crashes in jvm.dll in the compiler thread.

**Comment by KeesVo** [comment]
I have the same issue as well with 17.0.2+8:

---------------  S U M M A R Y ------------

Command Line: -Dorg.gradle.internal.worker.tmpdir=C:\Git\acp-hlc-emulation\Implementation\Projects\PROJ\impl\PROJMain\build\tmp\test\work -Dorg.gradle.native=false --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED -Xmx4G -Dfile.encoding=windows-1252 -Duser.country=US -Duser.language=en -Duser.variant -ea worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 1'

Host: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz, 12 cores, 31G,  Windows 10 , 64 bit Build 19041 (10.0.19041.1466)
Time: Mon Jan 31 09:58:48 2022 W. Europe Standard Time elapsed time: 56.646353 seconds (0d 0h 0m 56s)

---------------  T H R E A D  ---------------

Current thread (0x000001f8ff4cd340):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=4716, stack(0x000000299d900000,0x000000299da00000)]


Current CompileTask:
C2:  56646 21554       4       javafx.scene.control.TableView$TableViewArrayListSelectionModel::updateSelection (568 bytes)

Stack: [0x000000299d900000,0x000000299da00000],  sp=0x000000299d9fa350,  free space=1000k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [jvm.dll+0x57d9f2]
V  [jvm.dll+0x57d704]
V  [jvm.dll+0x57c978]
V  [jvm.dll+0x2202ed]
V  [jvm.dll+0x2206cf]
V  [jvm.dll+0x219f25]
V  [jvm.dll+0x2173eb]
V  [jvm.dll+0x1a3a8d]
V  [jvm.dll+0x226ef8]
V  [jvm.dll+0x22503c]
V  [jvm.dll+0x7e596b]
V  [jvm.dll+0x7dff2a]
V  [jvm.dll+0x6744d5]
C  [ucrtbase.dll+0x21bb2]
C  [KERNEL32.DLL+0x17034]
C  [ntdll.dll+0x52651]

**Comment by tkeksa** [comment]
It does not seem to be related to JavaFX code. See our crash below. Core dump is also available.

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffdc713d9f2, pid=11156, tid=32296
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.2+8 (17.0.2+8) (build 17.0.2+8)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.2+8 (17.0.2+8, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)
# Problematic frame:
# V  [jvm.dll+0x57d9f2]
#
# Core dump will be written. Default location: hs_err_pid11156.mdmp

---------------  S U M M A R Y ------------

Command Line: -Dlog4j.configurationFile=log4j2.xml V2XService.jar -c ./config/ -g ./config/appGeneralConfig.properties

Host: Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz, 4 cores, 11G,  Windows Server 2019 , 64 bit Build 17763 (10.0.17763.1339)
Time: Thu Feb 17 07:43:02 2022 Central Europe Standard Time elapsed time: 1431.802583 seconds (0d 0h 23m 51s)

---------------  T H R E A D  ---------------

Current thread (0x000002766bc28680):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=32296, stack(0x000000c5f0200000,0x000000c5f0300000)]


Current CompileTask:
C2:1431802 6625   !   4       ***<our package>***Stack::processPayloadWithTicket (183 bytes)

Stack: [0x000000c5f0200000,0x000000c5f0300000],  sp=0x000000c5f02fa920,  free space=1002k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [jvm.dll+0x57d9f2]
V  [jvm.dll+0x57d704]
V  [jvm.dll+0x57c978]
V  [jvm.dll+0x2202ed]
V  [jvm.dll+0x219ab9]
V  [jvm.dll+0x2173eb]
V  [jvm.dll+0x1a3a8d]
V  [jvm.dll+0x226ef8]
V  [jvm.dll+0x22503c]
V  [jvm.dll+0x7e596b]
V  [jvm.dll+0x7dff2a]
V  [jvm.dll+0x6744d5]
C  [ucrtbase.dll+0x21ffa]
C  [KERNEL32.DLL+0x17974]
C  [ntdll.dll+0x6a271]


siginfo: EXCEPTION_ACCESS_VIOLATION (0xc0000005), reading address 0x0000000000000008

**Comment by xupwup** [comment]
@karianna  This ticket has been on jbs:needs-report for a while, what does that mean? Should I report it at bugs.openjdk.java.net? And if so, where do I create an account?

**Comment by karianna** [comment]
It requires someone with Author rights to submit, I'll take a look for any duplicate bugs first.

**Comment by kingds** [comment]
Similar crash occurred in our application which uses Gluon JavaFX 16 on Temurin 17.0.1 (I suspect this is an uncommon pairing). We haven't been able to reliably reproduce. We hadn't seen this crash on OpenJDK 16 builds for what it's worth.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffdb936cce2, pid=9056, tid=13680
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.1+12 (17.0.1+12) (build 17.0.1+12)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (17.0.1+12, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)
# Problematic frame:
# V  [jvm.dll+0x57cce2]
#
# No core dump will be written. Minidumps are not enabled by default on client versions of Windows
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
#

---------------  S U M M A R Y ------------

Command Line: -Djpackage.app-version=6.3 --add-modules=javafx.controls,javafx.fxml,javafx.web --add-opens=javafx.base/com.sun.javafx.reflect=ALL-UNNAMED --add-opens=javafx.controls/javafx.scene.control.skin=ALL-UNNAMED --add-opens=javafx.graphics/com.sun.javafx.css=ALL-UNNAMED --add-opens=java.base/java.time=ALL-UNNAMED -DGENERA_HOME=C:\retisoft\genera\bin\app/../.. -Djpackage.app-path=C:\retisoft\genera\bin\Genera.exe com.retisoft.genera.gui.Main

Host: 11th Gen Intel(R) Core(TM) i5-11400 @ 2.60GHz, 12 cores, 63G,  Windows 10 , 64 bit Build 22000 (10.0.22000.434)
Time: Thu Feb 17 10:40:26 2022 Eastern Standard Time elapsed time: 67123.839703 seconds (0d 18h 38m 43s)

---------------  T H R E A D  ---------------

Current thread (0x00000151ff1892d0):  JavaThread "C2 CompilerThread0" daemon [_thread_in_native, id=13680, stack(0x00000014a1d00000,0x00000014a1e00000)]


Current CompileTask:
C2:67123839 25423       4       javafx.scene.control.TableView$TableViewArrayListSelectionModel::updateSelection (568 bytes)

Stack: [0x00000014a1d00000,0x00000014a1e00000],  sp=0x00000014a1dfa300,  free space=1000k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [jvm.dll+0x57cce2]
V  [jvm.dll+0x57c9f4]
V  [jvm.dll+0x57bc68]
V  [jvm.dll+0x22022d]
V  [jvm.dll+0x22060f]
V  [jvm.dll+0x219e6a]
V  [jvm.dll+0x217345]
V  [jvm.dll+0x1a355d]
V  [jvm.dll+0x226e38]
V  [jvm.dll+0x224f7c]
V  [jvm.dll+0x7e3f5b]
V  [jvm.dll+0x7de51a]
V  [jvm.dll+0x673305]
C  [ucrtbase.dll+0x26c0c]
C  [KERNEL32.DLL+0x154e0]
C  [ntdll.dll+0x485b]


siginfo: EXCEPTION_ACCESS_VIOLATION (0xc0000005), reading address 0x0000000000000008
```

**Comment by tkeksa** [comment]
I can provide few core dumps if it helps.

**Comment by karianna** [comment]
@tkeksa a core dump would be useful, thanks!

**Comment by tkeksa** [comment]
@karianna what is your preferred way how to send it to you? Compressed file has about 100 MB.

**Comment by karianna** [comment]
Can you zip upload it here?  Not sure what the GH limit is these days.

**Comment by kingds** [comment]
> Using -XX:+PrintCompilation we see that the specific method mentioned in the log (javafx.scene.control.TableView$TableViewArrayListSelectionModel::updateSelection) is compiled and invalidated several times, but we don't know how to trigger the exception.

@KeesVo could you please share how you configured this? I would like to try the same thing but I am unclear on how to set up the JVM args to point towards that method.

**Comment by kingds** [comment]
Hi @karianna any update on this? It seems like an OpenJDK issue is called for here.

**Comment by karianna** [comment]
> EXCEPTION_ACCESS_VIOLATION (0xc0000005)

For your case, please do try 17.0.2 and let us know the result.

**Comment by xupwup** [comment]
> It requires someone with Author rights to submit, I'll take a look for any duplicate bugs first.
Hi @karianna, did you find anything?

**Comment by karianna** [comment]
I couldn't find a duplicate no - added the report to JBS tag

**Comment by karianna** [comment]
https://bugs.openjdk.java.net/browse/JDK-8283565

**Comment by karianna** [comment]
I'll link it internally, but I suspect they are separate crashes, let's see!

**Comment by dean-long** [comment]
Yes, same crash.

**Comment by karianna** [comment]
@KeesVo If you have a file at $JAVA_HOME/bin/server/jvm.map in a Windows JDK, if it is included, can you post that here? 
 Another option would be to grab one of or fastdebug builds (check hte archives on the website) and reproducing with a "fastdebug" build can also be helpful, because the crash output will normally have more symbolic information, and the crash itself might be a more informative assert. (thanks @dean-long for the advice)

**Comment by karianna** [comment]
@KeesVo - Actually you don't need to worry, I read the upstream duplicate bug and they know where the issue is.

**Comment by xupwup** [comment]
@karianna, can you please let Tobias Hartmann (who's working on https://bugs.openjdk.org/browse/JDK-8283386) know that we (@KeesVo and me), can reproduce this issue consistently within 1 minute. Sadly we need our own proprietary software to do it, but perhaps we can arrange a meeting via MSTeams or something.

**Comment by TobiHartmann** [comment]
Thanks, let's follow-up by email: tobias.hartmann@oracle.com

**Comment by TobiHartmann** [comment]
We were able to narrow it down to [JDK-827561](https://bugs.openjdk.org/browse/JDK-8275610) which was fixed in JDK 11.0.15, 17.0.3 and 18.
